### PR TITLE
*: add mysql client test

### PR DESCRIPTION
**Pull request by xhebox**
### What problem does this PR solve?

Issue Number: close #xxx

Problem Summary:

What is changed and how it works:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Notable changes

- [ ] Has configuration change
- [ ] Has HTTP API interfaces change
- [ ] Has tiproxyctl change
- [ ] Other user behavior changes

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/proxy/backend/authenticator.go [hunk 1]
```diff
@@ -220,6 +223,13 @@ func (auth *Authenticator) handshakeFirstTime(logger *zap.Logger, cctx ConnConte
 		case mysql.ErrHeader:
 			return pnet.ParseErrorPacket(serverPkt)
 		default: // mysql.AuthSwitchRequest, ShaCommand
+			fmt.Printf("%d, %s, %+v\n", serverPkt[0], pluginName, serverPkt)
+			if serverPkt[0] == mysql.AuthSwitchRequest {
+				pluginName = string(serverPkt[1 : bytes.IndexByte(serverPkt[1:], 0)+1])
+			} else if serverPkt[0] == 1 && pluginName == "caching_sha2_password" && len(serverPkt) == 2 && serverPkt[1] == 3 {
+				// skip caching_sha2_password fast path
+				continue loop
+			}
```
Comment (djshow832) [inline pkg/proxy/backend/authenticator.go]: Why?

File: pkg/manager/namespace/manager.go [hunk 2]
```diff
@@ -33,7 +34,7 @@ func (mgr *NamespaceManager) buildNamespace(cfg *config.Namespace) (*Namespace,
 	logger := mgr.logger.With(zap.String("namespace", cfg.Namespace))
 
 	var fetcher router.BackendFetcher
-	if mgr.tpFetcher != nil {
+	if !reflect.ValueOf(mgr.tpFetcher).IsNil() {
```
Comment (djshow832) [inline pkg/manager/namespace/manager.go:37]: Why?

Comment (xhebox) [reply pkg/manager/namespace/manager.go:37]: Even if `pdFetcher` is nil, but when it is passing to a interface, the interface becomes a non-nil interface with nil value.

I mean `mgr.tpFetcher != nil` is always true.

File: pkg/proxy/backend/authenticator.go (hunk 1 above)
Comment (xhebox) [reply pkg/proxy/backend/authenticator.go]: There is a special fast path in `caching_sha2_password` where server sent a packet but no client resp.

File: .github/workflows/integration.yml [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        run: make cmd
+      - name: build mysql client test
+        run: |
+          cmake -Bbuild -DWITHOUT_SERVER=on -DWITH_UNIT_TESTS=off mysql-server
+          cmake --build build -t mysql_client_test
+      - name: setup & test
+        run: |
+          sed -i 's/# pd-addrs = "127.0.0.1:2379"/pd-addrs = ""/' conf/proxy.toml
+          sed -i 's/# auto-certs = true/auto-certs = true/' conf/proxy.toml
+          sed -i 's/4000/3306/' conf/namespace/example.toml
```
Comment (djshow832) [inline .github/workflows/integration.yml:60]: I suggest creating another config files in the action. Do not rely on the content in conf/proxy.toml.

Comment (xhebox) [reply .github/workflows/integration.yml:60]: Maybe later in tidb-test

File: pkg/proxy/backend/authenticator.go [hunk 4]
```diff
@@ -220,6 +223,12 @@ func (auth *Authenticator) handshakeFirstTime(logger *zap.Logger, cctx ConnConte
 		case mysql.ErrHeader:
 			return pnet.ParseErrorPacket(serverPkt)
 		default: // mysql.AuthSwitchRequest, ShaCommand
+			if serverPkt[0] == mysql.AuthSwitchRequest {
+				pluginName = string(serverPkt[1 : bytes.IndexByte(serverPkt[1:], 0)+1])
+			} else if serverPkt[0] == 1 && pluginName == "caching_sha2_password" && len(serverPkt) == 2 && serverPkt[1] == 3 {
+				// skip caching_sha2_password fast path
```
Comment (djshow832) [inline pkg/proxy/backend/authenticator.go:229]: I think the fast path is not skipped. Why do you comment "skip"?

File: pkg/proxy/backend/authenticator.go [hunk 5]
```diff
@@ -220,6 +223,12 @@ func (auth *Authenticator) handshakeFirstTime(logger *zap.Logger, cctx ConnConte
 		case mysql.ErrHeader:
 			return pnet.ParseErrorPacket(serverPkt)
 		default: // mysql.AuthSwitchRequest, ShaCommand
+			if serverPkt[0] == mysql.AuthSwitchRequest {
+				pluginName = string(serverPkt[1 : bytes.IndexByte(serverPkt[1:], 0)+1])
+			} else if serverPkt[0] == 1 && pluginName == "caching_sha2_password" && len(serverPkt) == 2 && serverPkt[1] == 3 {
```
Comment (djshow832) [inline pkg/proxy/backend/authenticator.go:228]:
```suggestion
			} else if serverPkt[0] == 1 && pluginName == mysql.AuthCachingSha2Password && len(serverPkt) == 2 && serverPkt[1] == 3 {
```

File: pkg/proxy/backend/authenticator.go (hunk 4 above)
Comment (xhebox) [reply pkg/proxy/backend/authenticator.go:229]: You can comment if you don't like. For me, it 's 'skip fast path (packets)`.

Comment (djshow832) [reply pkg/proxy/backend/authenticator.go:229]: Just comment "caching_sha2_password fast path".

**Review by djshow832 — APPROVED** [review]

**Pull request merged**